File: Solutions in Python/p231.py
# 
# Solution to Project Euler problem 231
# Copyright (c) Sankha Saha (CODR). All rights reserved.
# 
#
# https://github.com/codr07/Project-Euler-Solusions
# 

import eulerlib


def compute():
	N = 20000000
	K = 15000000
	smallestprimefactor = eulerlib.list_smallest_prime_factors(N)
	
	def factorial_prime_factor_sum(n):
		result = 0
		for i in range(n + 1):
			j = i
			while j > 1:
				p = smallestprimefactor[j]
				result += p
				j //= p
		return result
	
	ans = factorial_prime_factor_sum(N) - factorial_prime_factor_sum(K) - factorial_prime_factor_sum(N - K)
	return str(ans)


if __name__ == "__main__":
	print(compute())

File: codes/java/p231.java
/* 
 * Solution to Project Euler problem 231
 * Copyright (c) Sankha Saha. All rights reserved.
 * 
 *  
 *   
 */


public final class p231 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p231().run());
	}
	
	
	private static final int N = 20000000;
	private static final int K = 15000000;
	
	public String run() {
		smallestPrimeFactor = Library.listSmallestPrimeFactors(N);
		return Long.toString(factorialPrimeFactorSum(N)
			- factorialPrimeFactorSum(K)
			- factorialPrimeFactorSum(N - K));
	}
	
	
	private int[] smallestPrimeFactor;
	
	private long factorialPrimeFactorSum(int n) {
		long sum = 0;
		for (int i = 1; i <= n; i++) {
			int j = i;
			while (j > 1) {
				int p = smallestPrimeFactor[j];
				sum += p;
				j /= p;
			}
		}
		return sum;
	}
	
}
